### PR TITLE
[CALCITE-2889] IndexOutOfBoundsException thrown if targetTypes[] contains varargs

### PR DESCRIPTION
**Pull request by amaliujia**
target type could be variable length type (e.g String[].class), so the number of expressions might be great than length of target types, which leads to IndexOutOfBoundsException.

### CODE REVIEW

**Comment by zabetak** [comment]
Hi @amaliujia , can you add some tests which use a higher level API instead of directly calling EnumUtils#fromInternal?

**Comment by amaliujia** [comment]
@zabetak 

No problem. I had exception at  [ReflectiveCallNotNullImplementor.java#L48](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/adapter/enumerable/ReflectiveCallNotNullImplementor.java#L48). I can try to test it based on `ReflectiveCallNotNullImplementor`

**Comment by zabetak** [comment]
ReflectiveCallNotNullImplementor is still a bit low level. How did you end up there?

**Comment by amaliujia** [comment]
@zabetak 

In BeamSQL, we support UDF by [this way](https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/ScalarFunctionImpl.java#L122). We also use Calcite to generate java code for SQL functions (see [here](https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java#L149)).

Basically `ReflectiveCallNotNullImplementor` is the one to convert UDF in BeamSQL to Expression.

**Comment by zabetak** [comment]
Thanks for the explanation @amaliujia, now I see better! I suppose you can add an addtional test under UdfTest class.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java [hunk 1]
```diff
@@ -211,7 +211,23 @@ static Expression fromInternal(Expression e, Class<?> targetType) {
       List<Expression> expressions) {
     final List<Expression> list = new ArrayList<>();
     for (int i = 0; i < expressions.size(); i++) {
-      list.add(fromInternal(expressions.get(i), targetTypes[i]));
+      if (targetTypes[i].isArray()) {
+        // only the last type could be variable length argument. All left expressions
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java]: Only the last can be a vararg but there can be other array types in between, does this pose a problem? Maybe add some unit tests with such cases?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java [hunk 2]
```diff
@@ -211,7 +211,23 @@ static Expression fromInternal(Expression e, Class<?> targetType) {
...
+        list.addAll(fromInternal(targetTypes[i].getComponentType(), subList));
+        break;
+      } else {
+        list.add(fromInternal(expressions.get(i), targetTypes[i]));
+      }
+    }
+    return list;
+  }
+
+  static List<Expression> fromInternal(Class<?> targetType, List<Expression> expressions) {
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java]: Does it need to be package private? Can't we make it just private?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java (hunk 1 above)
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java]: Actually, this is the one comment that I care most. The rest are minor and I could treat them myself I think.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java (hunk 2 above)
Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java]: Updated

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java (hunk 1 above)
Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java]: Your concern is valid. I added a check to check if it's checking the last type to decide if move forward for varargs logic. 

Note that only the last type could be varargs. Complier guarantees it. Array types have values as arrays. So except for varargs, other types (including array types) are still 1:1 mapped to values.

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java]: Have added more tests for this case.

**Comment by amaliujia** [comment]
All green! @zabetak PTAL

**Comment by amaliujia** [comment]
Ping!

**Comment by amaliujia** [comment]
Ping!

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java [hunk 3]
```diff
@@ -202,7 +202,23 @@ static Expression fromInternal(Expression e, Class<?> targetType) {
       List<Expression> expressions) {
     final List<Expression> list = new ArrayList<>();
     for (int i = 0; i < expressions.size(); i++) {
-      list.add(fromInternal(expressions.get(i), targetTypes[i]));
+      if (i == targetTypes.length - 1 && targetTypes[i].isArray()) {
+        // only the last type could be varargs. All left expressions
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java]: Should we consider just `Array` type as var-length here ? Should we enumerate all the variable length types ?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java [hunk 4]
```diff
@@ -202,7 +202,23 @@ static Expression fromInternal(Expression e, Class<?> targetType) {
...
+        break;
+      } else {
+        list.add(fromInternal(expressions.get(i), targetTypes[i]));
+      }
+    }
+    return list;
+  }
+
+  private static List<Expression> fromInternal(Class<?> targetType, List<Expression> expressions) {
+    final List<Expression> list = new ArrayList<>();
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java]: Maybe we should rename the method cause there is already a method named `fromInternal`, say `fromComponentInternal` ?

File: core/src/test/java/org/apache/calcite/adapter/enumerable/EnumUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Tests for EnumUtilsTest.
+ */
+public class EnumUtilsTest {
+  @Test
+  public void testEnumUtilsFromInternalWithoutVarargsType() {
+    Class<?>[] targetTypes = new Class<?>[2];
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/adapter/enumerable/EnumUtilsTest.java]: Please keep the annotation same line with the method name.

File: core/src/test/java/org/apache/calcite/adapter/enumerable/EnumUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Tests for EnumUtilsTest.
+ */
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/adapter/enumerable/EnumUtilsTest.java]: Tests for EnumUtils.

**Comment by danny0405** [comment]
Besides of the unit test, we should add a ITCase also comes from the sql.

**Comment by zabetak** [comment]
> Besides of the unit test, we should add a ITCase also comes from the sql.

There is discussion in the JIRA with respect to this. I am not sure if it will be easy to reproduce but let's at least understand why.

**Comment by amaliujia** [comment]
Thanks for reviewing this PR. As I don't have enough bandwidth on it. I will close this PR and anyone who is interested could take this JIRA.

**Comment by zabetak** [comment]
Hi @amaliujia, I think the PR is close to being merged so I am re-opening it since I don't want you to lose the credit for your work so far. This does not mean that you are obliged to participate in the discussion or respond to the comments. Me or another reviewer may handle the things that are left. If it turns out that something big is missing we may close it in the future. Thanks again for working on this!

**Comment by amaliujia** [comment]
Per https://issues.apache.org/jira/browse/CALCITE-2889, this PR is half of feature. The full feature is support varargs in UDF. Will close this PR.
